### PR TITLE
EXISTS CIP

### PR DESCRIPTION
**Pull request by Mats-SX**
Supersedes #10.

### CODE REVIEW

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+Variables introduced in the subquery are not available on the outside.
+
+For each matching subgraph evaluated with `EXISTS {}`, the result value must be `true` if the subquery finds at least one matching row.
+If no matches are found, `false` should be returned.
+
+The subquery used in `EXISTS {}` has to follow the following rules:
+
+ * It has to be a read-only query. Updating the graph as part of a predicate is not allowed.
+ * It cannot end in a `RETURN` clause, nor may it contain `UNION`.
+ * If `WITH` ends the subquery, it should be able to change the cardinality of the subquery.
```
Comment (alastai) [inline cip/CIP2015-05-13-EXISTS.adoc]: This is test comment by Alastair

Comment (alastai) [inline cip/CIP2015-05-13-EXISTS.adoc]: This clause is redundant and contradicts the grammar definition of 

> read only clause

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+In the property checking form, any expression is accepted as the input.
+If the input expression evaluates to `null`, `exists()` will evaluate to `false`. For any other value, it will evaluate to `true`.
+This form of `exists()` is equivalent to checking for `null` using `IS NOT NULL`: `exists(expr)` is equivalent to `expr IS NOT NULL`.
+
+This form is there to make property existence checking idiomatic in Cypher - we do not want the user thinking of missing properties as checking for `null` values.
+
+==== Subquery exists
+
+When checking for the existence of subqueries, variables and parameters from the outer query are available in the subquery.
+Variables introduced in the subquery are not available on the outside.
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: Perhaps say something about this being a correlated subquery? We've done the same for the 'general' subqueries CIPed elsewhere...

File: cip/CIP2015-05-13-EXISTS.adoc [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+For each matching subgraph evaluated with `EXISTS {}`, the result value must be `true` if the subquery finds at least one matching row.
+If no matches are found, `false` should be returned.
+
+The subquery used in `EXISTS {}` has to follow the following rules:
+
+ * It has to be a read-only query. Updating the graph as part of a predicate is not allowed.
+ * It cannot end in a `RETURN` clause, nor may it contain `UNION`.
+ * If `WITH` ends the subquery, it should be able to change the cardinality of the subquery.
+ In other words, the `WITH` clause may include `WHERE` or `SKIP`, since these sub-clauses can turn a query that returns rows into one that does not.
+
```
Comment (petraselmer) [inline cip/CIP2015-05-13-EXISTS.adoc]: Is there a reason why EXISTS {...} cannot be nested? At the very least, this ought to be mentioned.

File: cip/CIP2015-05-13-EXISTS.adoc (hunk 2 above)
Comment (Mats-SX) [inline cip/CIP2015-05-13-EXISTS.adoc]: I'm not sure I understand what a 'correlated subquery' is. Do you have a reference?

File: cip/CIP2015-05-13-EXISTS.adoc (hunk 1 above)
Comment (Mats-SX) [inline cip/CIP2015-05-13-EXISTS.adoc]: I think that you might've misread this bullet point. It does state that `WITH` is allowed, and that is is further allowed to also change the cardinality of the subquery, through using `SKIP`, `LIMIT`, and most importantly `WHERE`. By changing cardinality like that, it does have the potential to alter the result of the subquery.

File: cip/CIP2015-05-13-EXISTS.adoc (hunk 3 above)
Comment (Mats-SX) [inline cip/CIP2015-05-13-EXISTS.adoc]: I don't think there is such a reason. I'll add a statement that makes it explicitly possible.

File: cip/CIP2015-05-13-EXISTS.adoc (hunk 1 above)
Comment (Mats-SX) [inline cip/CIP2015-05-13-EXISTS.adoc]: Perhaps the text is unintentionally easy to misunderstand? Perhaps we need to reword this?

Comment (Mats-SX) [inline cip/CIP2015-05-13-EXISTS.adoc]: Suggestion: 

If `WITH` ends the subquery, it is able to change the cardinality.

Comment (alastai) [inline cip/CIP2015-05-13-EXISTS.adoc]: You are right. If we change the words "should be able" to "is able" that would be clearer.

**Pull request merged**